File: 518_CoinChange2/518.cpp
//
// Created by hp on 2020/10/24.
//c++ version

#include <iostream>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    int change(int amount, vector<int>& coins) {
        vector<int>dp(amount+1,0);
        dp[0]=1;
        for(auto coin:coins){
            for (int i = coin; i <= amount; ++i) {
                dp[i] += dp[i-coin];
            }
        }
        return dp[amount];
    }
};

int main(){
    Solution a;
    vector<int> test ={1,2,5};
    int res = a.change(5,test);
    cout<<res<<endl;
}

File: 518_CoinChange2/518.java
//java version
package com.zhang.dp;
public class Solution_518 {
    public int change(int amount, int[] coins) {
        int[] dp = new int[amount+1];
        dp[0] = 1;
        for (int coin:coins)
            for (int i = coin; i <= amount; i++)
                dp[i] +=dp[i-coin];
        return dp[amount];
    }

    public static void main(String[] args) {
        Solution_518 solution_518 = new Solution_518();
        int[] amount = {1,2,5};
        int res = solution_518.change(5, amount);
        System.out.println(res);
    }
}